include/ginkgo/core/config/registry.hpp

```diff
@@ -0,0 +1,219 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2023, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_CONFIG_REGISTRY_HPP_
+#define GKO_PUBLIC_CORE_CONFIG_REGISTRY_HPP_
+
+
+#include <complex>
+#include <functional>
+#include <map>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils_helper.hpp>
+#include <ginkgo/core/config/property_tree.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+
+
+namespace gko {
+namespace config {
+
+
+class registry;
+
+
+using linop_map = std::unordered_map<std::string, std::shared_ptr<LinOp>>;
+using linopfactory_map =
+    std::unordered_map<std::string, std::shared_ptr<LinOpFactory>>;
+using criterionfactory_map =
+    std::unordered_map<std::string, std::shared_ptr<stop::CriterionFactory>>;
+using type_descriptor = std::pair<std::string, std::string>;
```

Comment:
I really don't think `string` is the right type that we should use here. We basically have a two stage dispatch: 
1. we determine the concrete index/value type we want to use at runtime from a mix of compile time choice (`parse<double, ...>`) and runtime choice (`'value_type': 'float64')
2. we pick the correct compile time type

I don't think `string` is a good option for this process. It decouples the runtime- and compile-time types. There is no simple way to get the valid compile-time types for a given runtime type string. Likewise, getting from a compile-time type to a runtime type involves some extra work that is done by macros here.

I think these issues are best addressed with `std::variant`. The valid compile-time choices are very clearly encoded, and creating the runtime type just uses the constructor.

This is the dispatch for 2. using variants:
https://github.com/ginkgo-project/ginkgo/blob/0f459b65535f12d0b43535a13d99ba53aca30adb/include/ginkgo/core/config/dispatch.hpp#L21-L47
Most of the actual dispatch is done by the `std::visit`, which will run the lambda and cast the variant to the stored type.

Here is the dispatch for 1:
https://github.com/ginkgo-project/ginkgo/blob/0f459b65535f12d0b43535a13d99ba53aca30adb/include/ginkgo/core/config/type_config.hpp#L42-L68
This uses the helper struct:
https://github.com/ginkgo-project/ginkgo/blob/0f459b65535f12d0b43535a13d99ba53aca30adb/include/ginkgo/core/config/type_config.hpp#L33-L39
which makes the connection between type and name used in the config quite simple. E.g. you can just have `alternative{"float32", float{}}`.

Reply:
the config can only give the string. We need to transfer the string to the compile time type at some point.
config(string) -> dispatch (std::string -> type)
config(string) -> variant (std::string -> type) -> dispatch
The connection part is also easy in the string by creating a mapping from the name to the standard name.
Currently, we only use standard c++14.

Reply:
I would argue that the `dispatch` using variants is also simpler than with strings. In the string case, you have no guarantee that the string value actually matches one of the map items. With the variant, it is guaranteed that the runtime value is exactly one of the valid compile-time types. This removes any consistency checks during the dispatch.

If you add a check that the runtime string matches one of the valid comile-time types, then you have exactly the same number of steps as the variant approach.

Also, how would you create the type mapping? I fear that it would again contain a lot of macro magic.  But since we have an option without macros, I would really prefer that one.

TBH, I feel like the only argument for the string approach is that it's c++14. But I think that is a very weak argument. The variant approach requires only library features of c++17, but not language features. So we can just take a c++14 implementation of `std::variant`, which I already mentioned in #1389.

Reply:
the check is in the dispatch part. It always needs the check `if (str == string(type)) -> call type`. The unmatched pair will go to the end implementation. The number of steps is not important, but I mean both of them can only accept the allowed runtime -> compiled type up to the different place to verify it.

The type mapping is 
```
template <>
struct type_string<double> {
    static std::string str() {return "double";}
};
```
`type_string<type>::str();`  get the type string.

Yes, the main point is c++14 without any third-party dependence. I usually try to avoid third party if it is an easy implementation.

Reply:
We copied whole parts of SuiteSparse into our code. So I think we already allow including third-party code if it's beneficial for the implementation (which I suggest it is). Also, that third-party code should not stay for long, since we want to update to c++17 anyway.